### ISSUE TITLE
[Question] WAF, Ingress Controller and API Gateway. Are they all needed?

### ISSUE THREAD

**Issue by jooooel** [dotnet/eShopOnContainers#1818]
Hi!
When deploying this to Kubernetes, NGINX is used as an Ingress Controller. Envoy is used internally in the cluster as an API Gateway. And there are [mentions of using Azure Application Gateway](https://github.com/dotnet-architecture/eShopOnContainers/issues/752#issuecomment-418951853).

Would it be possible to skip at least one of them and handle it all in the Ingress Controller? 

I guess my concern is that having three separate components that (possibly) can handle all of it, adds extra complexity and latency. I know that Azure Application Gateway can be used as an Ingress Controller for example (however, I've seen some issues related to downtime while deploying configuration that doesn't sound like a good option). And Envoy can also be used as an Ingress Controller as well.

The reason I'm asking, is because I'm facing a similar issue. We are using Traefik as an Ingress Controller for ans AKS cluster, and we are looking at adding an API Gateway inside the cluster, but also some kind of WAF external to the cluster. I'm trying to figure out if they all need to be three separate pieces or if they can (and should) be handled by a single or maybe two components instead.

The requirements in my case:
- Authentication
- Routing
- TLS termination (and automatic certificate management)
- Security features similar to what a WAF would provide
- Something that works well with local development as well

**Comment by erjain** [comment]
Hi @jooooel, we used Envoy for internal routing because when we moved away from Ocelot, we found the Envoy to support gRPC, and our service mesh requirements quite nicely. However, since our needs are minimal for this sample, NGINX-based routing was more than enough, and we made that choice while we built this module https://docs.microsoft.com/en-us/learn/modules/microservices-apigateway-aspnet-core/. We also recommended Azure Application Gateway with AGIC for production scenarios in the same module for the same considerations you mentioned here. If you are facing issues with AGIC, we recommend reaching out to the teams directly by reporting an issue here: https://github.com/Azure/application-gateway-kubernetes-ingress/issues. Hope this will help you.
